packages/ai-jsx/src/batteries/sidekick/platform/large-response-handler.tsx

```diff
@@ -0,0 +1,222 @@
+
+  // Option 1: do nothing if it's already small enough
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  stringified = yamlOptimizeIfPossible(stringified);
+
+  // Option 2: try dumping as YAML. If it's small enough, then we are done.
+  if ((await lengthFunction(stringified)) <= maxLength) {
+    return <FunctionResponse {...props}>{stringified}</FunctionResponse>;
+  }
+
+  // Option 3 (last reosrt): split into chunks and allow LLM to query by similarity
+  // Requires Cohere API key for doing similarity search
+  const cohereConfig = getContext(cohereContext);
+
+  if (!cohereConfig.api_key) {
+    // Not yet an error, but this is an error just waiting to happen
+    logger.warn(
+      { CohereContext: cohereConfig },
+      'FunctionResponse is too big, but Cohere API key is not set. Please set it in the context.'
+    );
+  }
+
+  const splitter = new RecursiveCharacterTextSplitter({
+    chunkSize: maxLength / numChunks,
+    chunkOverlap: maxLength / numChunks / 10,
+    lengthFunction,
+  });
+  const chunks = await splitter.splitText(stringified);
+
+  return (
+    <FunctionResponse {...props} metadata={{ isRedacted: true, chunks }}>
+      ... The response is too big and hence redacted. The response can be queried using semantic similarity search by
+      calling the `loadBySimilarity` function.
+    </FunctionResponse>
+  );
+  // TODO: return a chunk based on query (gotta get it from conversation)
+}
+
+/**
+ * This function allows wrapping {@link FunctionResponse} elements that can possibly be too large.
+ * It will replace FunctionResponse elements with {@link LargeFunctionResponseHandler}s that know how to handle large responses.
+ *
+ * {@link LargeFunctionResponseHandler} will not modify responses that are not large. If they are, it will first try
+ * to optimize the response by dumping it as YAML, which is more token-efficient than JSON.
+ * If that doesn't work, it will split the response into chunks and allows the LLM to query it using semantic similarity
+ * search by exposing a dynamic function called `loadBySimilarity(query)`.
+ *
+ * Note that failed responses are not optimized and will simply be truncated to `failedMaxLength` characters.
+ *
+ * @see {@link redactedFunctionTools} to see how to add the `loadBySimilarity` function.
+ *
+ * @example
+ * ```tsx
+ *    <LargeFunctionResponseWrapper maxLength={4000} failedMaxLength={1000} numChunks={4}>
+ *      <ExecuteFunction func={tool[name].func} name={name} args={args} />
+ *    </LargeFunctionResponseWrapper>
+ * ```
+ */
+export async function LargeFunctionResponseWrapper(
+  { children, ...props }: { children: AI.Node } & LargeFunctionResponseProps,
+  { render }: AI.ComponentContext
+) {
+  // We need to render the children to get the FunctionResponse elements
+  const elements = await render(children, { stop: (e) => e.tag == FunctionResponse });
```

Comment:
I think using `renderToConversation` and then replacing just the `<FunctionResponse>`s  is preferable. Would rather minimize the proliferation of partial rendering.